homeassistant/components/zwave_js/__init__.py

```diff
@@ -36,69 +38,130 @@ def register_node_in_dev_reg(
+        LOGGER.debug("Processing node %s", node)
+
+        # register (or update) node in device registry
+        register_node_in_dev_reg(entry, dev_reg, client, node)
+
+        # run discovery on all node values and create/update entities
+        for disc_info in async_discover_values(node):
+            LOGGER.debug("Discovered entity: %s", disc_info)
+            async_dispatcher_send(hass, f"{DOMAIN}_add_{disc_info.platform}", disc_info)
+
+    @callback
+    def async_on_node_added(node: ZwaveNode) -> None:
+        """Handle node added event."""
+        LOGGER.debug("Node added: %s - waiting for it to become ready.", node.node_id)
+        # we only want to run discovery when the node has reached ready state,
+        # otherwise we'll have all kinds of missing info issues.
+        if node.ready:
+            async_on_node_ready(node)
+            return
+        # if node is not yet ready, register one-time callback for ready state
+        node.once(
+            "ready",
+            lambda event: async_on_node_ready(event["node"]),
         )
+        # we do submit the node to device registry so user has
+        # some visual feedback that something is (in the process of) being added
+        register_node_in_dev_reg(entry, dev_reg, client, node)
+
+    async def handle_ha_shutdown(event):
+        """Handle HA shutdown."""
+        await client.disconnect()
 
     # register main event callbacks.
     unsubs = [
         client.register_on_initialized(async_on_initialized),
         client.register_on_disconnect(async_on_disconnect),
+        client.register_on_connect(async_on_connect),
+        hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, handle_ha_shutdown),
     ]
 
+    # connect and throw error if connection failed
     asyncio.create_task(client.connect())
-
     try:
         async with timeout(10):
             await initialized.wait()
     except asyncio.TimeoutError as err:
-        for unsub in unsubs:
+        for unsub in hass.data[DOMAIN][entry.entry_id][DATA_UNSUBSCRIBE]:
             unsub()
         await client.disconnect()
         raise ConfigEntryNotReady from err
 
-    async def handle_ha_shutdown(event):
-        """Handle HA shutdown."""
-        await client.disconnect()
+    platforms_loaded = 0
+    platforms_ready = asyncio.Event()
 
-    unsubs.append(hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, handle_ha_shutdown))
+    @callback
+    def mark_platform_ready():
+        nonlocal platforms_loaded
+        platforms_loaded += 1
+        if len(PLATFORMS) == platforms_loaded:
+            platforms_ready.set()
 
     hass.data[DOMAIN][entry.entry_id] = {
-        "client": client,
-        "unsubs": unsubs,
+        DATA_CLIENT: client,
+        DATA_UNSUBSCRIBE: unsubs,
+        DATA_PLATFORM_READY: mark_platform_ready,
     }
 
+    async def when_platforms_ready():
+        """Call when all supported/required platforms are ready."""
+        await platforms_ready.wait()
```

Comment:
Why don't we await the gather of platform forwarding here instead? I don't think we need to have the event and count platforms. The forwarding task will not be done until the platform setup task is done.

Reply:
That was there and didn't work.

Reply:
- Awaiting platform fowards blocks entire setup.
- We want to check if the connection works or else raise ConfigEntryNotReady
- discovery may only be run after platforms are setup AND client connected

this was the race condition I was talking about with @balloob and he changed the code to this which actually covers all usecases and works.

Reply:
hmmm, but we could start a task that awaits the platforms...

Reply:
![Schermafbeelding 2021-01-08 om 14 16 29](https://user-images.githubusercontent.com/6389780/104019591-1ac8c980-51bc-11eb-8604-6678a402b438.png)
this works too, without the event

Reply:
That was what I meant. 👍

Reply:
OK, I'll adjust it.